File: ios/Bridge/Hybrid/BarcodeScanningResult+toNitro.swift
#if MLKIT_BARCODE_SCANNING
  extension BarcodeScanningResult {
    func toNitroResult() throws -> BarcodeScannerResult {
      return BarcodeScannerResult(
        barcodes: try barcodes.map { try $0.toNitroBarcode() }
      )
    }
  }

  extension BarcodeData {
    fileprivate func toNitroBarcode() throws -> BarcodeScannerBarcode {
      return BarcodeScannerBarcode(
        bounds: bounds?.toNitroBoundingBox(),
        corners: corners?.map { $0.toNitroCorner() },
        format: Double(format),
        formatName: BarcodeFormat(fromString: formatName) ?? .unknown,
        valueType: Double(valueType),
        valueTypeName: BarcodeValueTypeName(fromString: valueTypeName) ?? .typeUnknown,
        rawValue: rawValue,
        displayValue: displayValue,
        rawBytes: rawBytes?.map { Double($0) },
        value: try value?.toAnyMap(),
        isPotential: isPotential
      )
    }
  }
#endif

File: android/src/main/java/com/margelo/nitro/visioncameramlkit/BarcodeScanningResult+toNitro.kt
package com.margelo.nitro.visioncameramlkit

internal fun com.visioncameramlkit.domain.models.BarcodeScanningResult.toNitroResult(): BarcodeScannerResult =
  BarcodeScannerResult(
    barcodes = barcodes.map { it.toNitroBarcode() }.toTypedArray(),
  )

private fun com.visioncameramlkit.domain.models.BarcodeData.toNitroBarcode(): BarcodeScannerBarcode =
  BarcodeScannerBarcode(
    bounds = bounds?.toNitroBoundingBox(),
    corners = corners?.toNitroCorners(),
    format = format.toDouble(),
    formatName = formatName.toNitroBarcodeFormat(),
    valueType = valueType.toDouble(),
    valueTypeName = valueTypeName.toNitroBarcodeValueTypeName(),
    rawValue = rawValue,
    displayValue = displayValue,
    rawBytes = rawBytes?.map { it.toDouble() }?.toDoubleArray(),
    isPotential = isPotential,
  )

private fun String.toNitroBarcodeFormat(): BarcodeFormat =
  runCatching { BarcodeFormat.valueOf(this) }.getOrDefault(BarcodeFormat.UNKNOWN)

private fun String.toNitroBarcodeValueTypeName(): BarcodeValueTypeName =
  runCatching { BarcodeValueTypeName.valueOf(this) }.getOrDefault(BarcodeValueTypeName.TYPE_UNKNOWN)
